File: 2020/python/day7.py
rules = {}
with open('../input/input7.txt', 'r') as fd:
    for rule in [x.strip().split(' ') for x in fd]:
        parent = "-".join(rule[0:2])
        rules[parent] = []
        if rule[4] != 'no':
            for i in range(4, len(rule), 4):
                rules[parent].append([int(rule[i]), "-".join(rule[i+1:i+3])])

def search(node, target):        
    for _, bag in rules[node]:
        if bag == target or search(bag, target):
            return True        
    return False

def count(node):    
    return 1 + sum(map(lambda x: x[0] * count(x[1]), rules[node]))

target = 'shiny-gold'
print("A:", sum(map(int, [search(x, target) for x in rules.keys() if x != target])))
print("B:", count(target) - 1)

File: 2020/haskell/day7.hs
import Data.Map (Map, fromList, keys, lookup)
import Data.Maybe (isJust, fromMaybe)
import Data.List (find)

bagId x = foldl1 (\a x -> a ++ "-" ++ x) (take 2 x)

bagList :: [String] -> [(Int, String)]
bagList [] = []
bagList xs 
    | head xs == "no" = []
    | otherwise = (n, id) : bagList (drop 4 xs)
        where 
            n = read (head xs) :: Int
            id = bagId (drop 1 xs)

parse :: [String] -> (String, [(Int, String)])
parse xs = (bagId xs, bagList rest)
    where 
        rest = drop 4 xs
search :: Map String [(Int, String)] -> String -> String -> Bool
search dict target x 
    | target == x = True
    | otherwise   = case Data.Map.lookup x dict of
        Just content -> isJust $ find (search dict target) (map snd content)
        Nothing      -> False

countBags :: Map String [(Int, String)] -> String -> Int
countBags dict x = 1 + sum (map (\x -> fst x * countBags dict (snd x)) bags)
    where
        bags = fromMaybe [] (Data.Map.lookup x dict)

main = do
    c <- readFile "../input/input7.txt"
    let target = "shiny-gold"
    let dict = fromList $ map (parse . words) (lines c)
    let partA = sum [fromEnum (search dict target x) | x <- keys dict, x /= target]
    putStrLn ("A: " ++ show partA)
    let partB = countBags dict target - 1
    putStrLn ("B: " ++ show partB)